### PR TITLE
Implementing VF instance selector

### PR DESCRIPTION
**Pull request by kontur**
Here's a start for #25 — I'm pretty sure there is things I missed with the overall flow of the app and with vanilla in particular, so happy to take your pointers.

I've added the instance parsing to otfFont.py for now — I'm not using the designspace workflow myself, but I assume it should expose them as well? I'm adding the family to the instance name, otherwise it becomes confusing with several files. I've opted for a default "No instance selected" and using the sliders also resets to this. 

One visual thing I couldn't figure out is how to give the PopUp the same margin as the slider container; I tried with those same "margin" settings and negative widths, but that just created havoc. Now the instance Popup goes from edge to edge, when it would be nicer to have that little margin. The PopUp is placed under the axes and I'm updating the y position inside the tab based off the axis widget's calculated height — not sure if there is a more automatic way to stack (and update) items inside a Vanilla Group.

### CODE REVIEW

File: Lib/fontgoggles/mac/mainWindow.py [hunk 1]
```diff
@@ -975,13 +982,21 @@ def featuresChanged(self, sender):
...
         self.textEntryChangedCallback(self.textEntry, updateCharacterList=False)
+        self.variationsGroup.instances.setItems([i[0] for i in allInstances])
+        # Reset instance popup
+        self.variationsGroup.instances.set(0)
 
     @objc.python_method
     def varInstanceChanged(self, sender):
-        self.project.textSettings.varLocation = {k: v for k, v in sender.get().items() if v is not None}
+        """Instance was selected from popup"""
+        _, _, _, allInstances, _, _ = self._gatherSidebarInfo(self.project.fonts)
```
Comment (kontur) [inline Lib/fontgoggles/mac/mainWindow.py:996]: These calls to get the `allInstances` are quite clumsy, but I wanted to keep the generation of the list in one spot to avoid confusion. Maybe inside `_gatherSidebarInfo` the instance part could be refactored into a separate method, and in this callback in `varLocationChanged` that separate method would suffice.

File: Lib/fontgoggles/mac/mainWindow.py [hunk 2]
```diff
@@ -559,13 +561,15 @@ def _gatherSidebarInfo(fonts):
             allAxes.append(font.axes)
             allScriptsAndLanguages.append(font.scripts)
             allStylisticSetNames.append(font.stylisticSetNames)
+            allInstances.extend(font.instances)
```
Comment (kontur) [inline Lib/fontgoggles/mac/mainWindow.py:564]: Just re-reading this... if there is the same VF several times in the list of project fonts this will presumably add the instances several times as well... should to a merge down of the list.

File: Lib/fontgoggles/font/baseFont.py [hunk 3]
```diff
@@ -92,6 +93,23 @@ def axes(self):
...
+    @cachedProperty
+    def instances(self):
+        fvar = self.ttFont.get("fvar")
+        if fvar is None:
+            return []
+        name = self.ttFont["name"]
+        instances = []
+        for i in fvar.instances:
+            try:
+                # Assuming VFs have ID 16 set; should this fallback to ID 1 or "" if not?
```
Comment (kontur) [inline Lib/fontgoggles/font/baseFont.py:105]: What's your thoughts on this?

File: Lib/fontgoggles/mac/mainWindow.py [hunk 4]
```diff
@@ -303,9 +303,10 @@ def setupSidebarGroup(self):
                                                        showHiddenAxes=self.project.uiSettings.showHiddenAxes)
 
         self.variationsGroup.instances = LabeledView(
-            (0, 0, 0, 1),  # gets updated as axes are read
+            # Position and instances gets updated as fonts are read
+            (0, 0, 0, 20),
```
Comment (kontur) [inline Lib/fontgoggles/mac/mainWindow.py:307]: What does the `x, y, w, h` destructuring in LabledView do, if only checking for `h > 0`  — should it be removed?

**Comment by justvanrossum** [comment]
Thanks! I'll need some time to review and advise. Please ping me here if you think it's taking too long.

Btw. what did you think of my suggestion in [#25](https://github.com/justvanrossum/fontgoggles/issues/25#issuecomment-1347908926) about an alternative approach to show named instances? This bit:

> An alternative may be to treat named instances like fonts in a collections (TTC). For example, we could add a contextual menu named "Expand named instances", and it would add all named instances as individual "fonts".

**Comment by kontur** [comment]
Sure, no problem, I might see if I can improve some of the things I commented on in the code above in the meantime.

Re. expanding the VF instances as pseudo fonts in the view: I think it is a different paradigm; the sliders and the instance selector work together, allowing the user to see what axis value an instance has (and somehow exposing _what instances there are_). Both are ambivalent in the sense that they affect to all fonts in the list if the font has such axes. 

I could imagine dropping a VF on the UI could prompt if to add as single VF or expand to all instances, as separate feature. An alternative could be right-clicking a VF in the list and it has all instances available for switching to.

In a nutshell both approaches have the same issue with being ambivalent... a choice relevant for one font possibly affects several fonts, or it is unclear why some fonts are not affected. I don't have a clear solution to offer for that, unfortunately.

File: Lib/fontgoggles/font/baseFont.py (hunk 3 above)
Comment (kontur) [reply Lib/fontgoggles/font/baseFont.py:105]: That is to say, should a VF have instances but is missing ID 16 it's instances will be added as "None - Regular Etc.".

Comment (kontur) [reply Lib/fontgoggles/font/baseFont.py]: Improved this to show _some_ family, always, as part of supporting designspaces.

File: Lib/fontgoggles/mac/mainWindow.py (hunk 2 above)
Comment (kontur) [reply Lib/fontgoggles/mac/mainWindow.py]: Improved this. I think the Popup widget had an automatic safeguard against identical items, but their order would be odd if adding the same fonts with intermitted other fonts. So unique list now, in order of instances read from fonts from the font list.

File: Lib/fontgoggles/mac/mainWindow.py (hunk 1 above)
Comment (kontur) [reply Lib/fontgoggles/mac/mainWindow.py:996]: I was wondering if all of what `_gatherSidebarInfo` returns shouldn't rather be e.g. an `@property` of `Project`. Then `_updateSidebarItems` could just access the properties as needed, and these callbacks could access the instance list as needed. The getters could cache the iteration over the font list internally, to avoid iterating the font list for each property when called without font list changes between calls to different (or same, for that matter) attributes.

File: Lib/fontgoggles/mac/mainWindow.py (hunk 4 above)
Comment (kontur) [reply Lib/fontgoggles/mac/mainWindow.py:307]: I really don't understand the `posSize` ... dug down into `VanillaBase` but can't figure it out. I tried `(10, y, -10, 0)` (updated also in the `_updateSidebarItems` call) like the sliders, but that just results in weird overlaps and cut offs.

**Comment by kontur** [comment]
Supports parsing the instance names from designspaces now as well. There is no proper name table entries for such a constructed font, so I am using the font list label as family fallback of the instance names.

**Comment by kontur** [comment]
This would still be handy to have, I constantly find myself launching the app from my fork instead of the installed version, exactly because I need this feature :)

I'd really only need help with the nested label positioning for which I cannot figure out how to give it some padding:
Screenshot 2023-02-27 at 9 51 40

**Comment by kontur** [comment]
> I'd really only need help with the nested label positioning for which I cannot figure out how to give it some padding

Any pointers? ☝️

**Comment by kontur** [comment]
Hmm. Has something changed with the dev flow? I can build the app, open it, but I can't open any files with it and the main window never opens (also on `master`).

```
Executing task: python App/setup.py py2app -A && ./App/dist/FontGoggles.app/Contents/MacOS/FontGoggles 

Hello .zprofile
/Users/johannes/Projects/fontgoggles/App/setup.py:2: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
running py2app
creating /Users/johannes/Projects/fontgoggles/App/build/bdist.macosx-12.2-arm64/python3.10-standalone/app
creating /Users/johannes/Projects/fontgoggles/App/build/bdist.macosx-12.2-arm64/python3.10-standalone/app/collect
creating /Users/johannes/Projects/fontgoggles/App/build/bdist.macosx-12.2-arm64/python3.10-standalone/app/temp
creating build/bdist.macosx-12.2-arm64/python3.10-standalone/app/lib-dynload
creating build/bdist.macosx-12.2-arm64/python3.10-standalone/app/Frameworks
Add paths for VENV /Users/johannes/.pyenv/versions/3.10.4 False
*** creating application bundle: FontGoggles ***
Copy '/Users/johannes/.pyenv/versions/fontgoggles/lib/python3.10/site-packages/py2app/apptemplate/prebuilt/main-arm64' -> '/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app/Contents/MacOS/FontGoggles'
sign ['/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app/Contents/MacOS/FontGoggles']
/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app/Contents/MacOS/FontGoggles: replacing existing signature
/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app/Contents/MacOS/FontGoggles: signed app bundle with Mach-O thin (arm64) [com.github.justvanrossum.FontGoggles]
/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app: replacing existing signature
/Users/johannes/Projects/fontgoggles/App/dist/FontGoggles.app: signed app bundle with Mach-O thin (arm64) [com.github.justvanrossum.FontGoggles]
Done!
/Users/johannes/Projects/fontgoggles/App/FontGoggles.py:11: DeprecationWarning: Set objc.options.verbose instead
  objc.setVerbose(True)
```

**Comment by justvanrossum** [comment]
There must be more errors once you try to open the app or try to open a window.

First thing to check: are all dependencies up to date? Perhaps try to create a new venv from scratch.

I'm currently building with Python 3.12, but it should still work on 3.10.

**Comment by kontur** [comment]
No errors, just no main window. I'm debugging myself along through the window instantiation with traces to see if something silently fails, but not found anything off yet.

https://github.com/justvanrossum/fontgoggles/assets/1893577/04427a30-37e2-492e-99d6-702832b5da5f

**Comment by justvanrossum** [comment]
Hmm, can you try with pyobjc==10.2 instead of the current 10.3? That init thing may be a clue.

**Comment by justvanrossum** [comment]
Yup, that's the one, I'll file a pyobc bug :( We need to pin to 10.2 for now.

File: Lib/fontgoggles/mac/mainWindow.py [hunk 5]
```diff
@@ -959,8 +982,21 @@ def featuresChanged(self, sender):
 
     @objc.python_method
     def varLocationChanged(self, sender):
+        """Axis slider value changed"""
+        _, _, _, allInstances, _, _ = self._gatherSidebarInfo(self.project.fonts)
         self.project.textSettings.varLocation = {k: v for k, v in sender.get().items() if v is not None}
         self.textEntryChangedCallback(self.textEntry, updateCharacterList=False)
+        self.variationsGroup.instances.setItems([i[0] for i in allInstances])
```
[the remark this answers is not in the document]
Comment (kontur) [reply Lib/fontgoggles/mac/mainWindow.py]: Left over from when I tried to select instances if matched; implemented in the latest commit.

File: Lib/fontgoggles/mac/mainWindow.py [hunk 6]
```diff
@@ -287,19 +287,35 @@ def setupSidebarGroup(self):
...
         self.featuresGroup = FeatureTagGroup(sidebarWidth - 6, {}, callback=self.featuresChanged)
         featuresTab.main = AligningScrollView((0, 0, 0, 0), self.featuresGroup, drawBackground=False,
                                               hasHorizontalScroller=False, autohidesScrollers=True,
                                               borderType=AppKit.NSNoBorder)
 
+        # Sidebar second tab
         variationsTab = group.feaVarTabs[1]
-        self.variationsGroup = SliderGroup(sidebarWidth - 6, {}, callback=self.varLocationChanged,
-                                           showHiddenAxes=self.project.uiSettings.showHiddenAxes)
+        self.variationsGroup = Group((0, 0, 0, 0))
```
[the remark this answers is not in the document]
Comment (kontur) [reply Lib/fontgoggles/mac/mainWindow.py:301]: I couldn't really get this to work like this because SliderGroup seemed to always place its nested items absolutely. I refactored how `LabledView` passes its `posSize` to child elements and amended this where it is used elsewhere (the top of the sidebar Orientation/Script/Language labels) — then I was able to properly offset the `variationsGroup` as intended.

**Comment by kontur** [comment]
> Yup, that's the one, I'll file a pyobc bug :( We need to pin to 10.2 for now.

And this is indeed needed, but seemed unrelated to this PR in particular so I left that out.

**Comment by justvanrossum** [comment]
> > Yup, that's the one, I'll file a pyobc bug :( We need to pin to 10.2 for now.

> And this is indeed needed, but seemed unrelated to this PR in particular so I left that out.

I just committed a pin to 10.2 to master, perhaps you want to rebase.
